test/config/utility.cc

```diff
@@ -113,6 +113,84 @@ name: envoy.squash
     nanos: 0
 )EOF";
 
+std::string ConfigHelper::discoveredClustersBootstrap(const std::string& api_type) {
```

Comment:
Should we also merge this into https://github.com/envoyproxy/envoy/blob/master/test/config/utility.h while we're here and make this more closely aligned with how the bootstraps are generated for the rest of the Envoy integration tests? CC @alyssawilk

Reply:
This function is already declared in utility.h. By "make this more closely aligned" are you referring to the selectable api_type being formatted in?

Reply:
I guess what I'm after here is something closer to how `ConfigHelper` works today. We have a very minimal bootstrap and then build it up programatically. In this YAML fragment, we have something that works well for a few tests today, but is less flexible, as it dictate the structure of the test, e.g. number of listeners.

Reply:
Yeah, while I'll delegate to Harvey how much we want to push on doing this now or later, I'd prefer we leverage existing integration test set up as much as we can.  If we only think we're going to want to test CDS in isolation, then we can have a hard-coded bootstrap here.  If someone's going to want to test the whole bootstrap flow from LDS to soon-to-be-FDS to CDS and EDS, I think we'd want to use the existing test set up and extend it to be more flexible, rather than one-off a set-up here which won't play well with others.

I'm hoping by end of year the integration test framework will just let you live load any component as we can in prod, and while getting things in one file helps, doing bootstrap from scratch doesn't

Reply:
Being able to have things happening with multiple xDS types in one test does sound good, and what you're describing would be necessary for that. However, wouldn't it also remain useful to preserve tests like this of the individual xDS types? With an isolated test like this, you don't have dynamic loading of other resource types potentially obfuscating things. (In fact, I'm thinking the hardcoded bootstrap config would make more sense living back in cds_integration_test.cc).

Reply:
If we can, I think I'd like to have config booleans for static vs dynamic for all the DSes.  So if they were all false, we'd get what we have today, and if they were all true we'd have what is described above, and whatever section is true has the hard-coded configuration swapped out for the discovery service config.

I guess what I'm wondering is if we care to put the work in now so we can just take the existing all-static config and swap the CDS information to be dynamic, rather than having a one-off config for CDS which writes everything from scratch and so can't be reused.

Reply:
Yeah, I'm inclined towards what @alyssawilk is suggesting if possible. I agree that that directionally, we want to go towards making everything programmatic. Basically, from my perspective, every time we have YAML boilerplate, vs. programmatic integration test bootstrap construction, we add technical debt, since we have an integration test that behaves subtley different to the rest, and where if we want to add some integration test-wide additional configuration we have an extra special case (that was a long sentence).

Reply:
Yes, the mix and match static/dynamic does sound like a great idea. It also sounds like it, or anything else in that direction, would be a significant change, beyond even the fact that there's not yet anything to base such changes on, since there is currently only a test for CDS. So, maybe this discussion could move from this PR to an issue? The existing repo has hardcoded YAML for the CDS test already; it's not like this PR is introducing it. In fact, there are 2, and this PR reduces it to 1.

Reply:
Sure, since we're in agreement, can you file an issue to track the extended work, add appropriate TODOs and then we can get this current merged?

Reply:
Done, #6327